File: lib/entity/article.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/cupertino.dart';

class Article {
  String content;
  String signature;
  num left;
  num top;
  num width;
  DateTime createdAt;
  String createdBy;

  bool get isPhoto {
    return content.contains('<img');
  }

  Widget? get photoImage {
    if (!isPhoto) {
      return null;
    }
    final pattern = RegExp('<img src="(.+)" data-rotation="(.+)"');
    final matches = pattern.allMatches(content);
    if (matches.isEmpty) {
      return null;
    }
    final src = matches.single.group(1);
    if (src == null) {
      return null;
    }
    return Image.network(src);
  }

  double? get rotation {
    if (!isPhoto) {
      return null;
    }
    final pattern = RegExp('<img src="(.+)" data-rotation="(.+)"');
    final matches = pattern.allMatches(content);
    if (matches.isEmpty) {
      return null;
    }
    final rotation = matches.single.group(2);
    if (rotation == null) {
      return null;
    }
    return double.parse(rotation);
  }

  Article({
    required this.content,
    required this.signature,
    required this.left,
    required this.top,
    required this.width,
    required this.createdAt,
    required this.createdBy,
  });

  factory Article.fromSnapShot(DocumentSnapshot snapshot) {
    return Article(
      content: snapshot.get('content'),
      signature: snapshot.get('signature'),
      left: snapshot.get('left'),
      top: snapshot.get('top'),
      width: snapshot.get('width'),
      createdAt: (snapshot.get('createdAt') as Timestamp).toDate(),
      createdBy: snapshot.get('createdBy'),
    );
  }

  toJson() {
    return {
      "content": content,
      "signature": signature,
      "left": left,
      "top": top,
      "width": width,
      "createdAt": Timestamp.fromDate(createdAt),
      "createdBy": createdBy,
    };
  }
}

File: types/article.ts
import { Timestamp, type DocumentData, type QueryDocumentSnapshot } from "firebase/firestore";

export type Article = {
  id: string;
  content: string;
  signature: string;
  left: number;
  top: number;
  width: number;
  createdAt: Date;
  createdBy: string;
};

export type ArticleInput = Omit<Article, "id" | "createdAt"> & {
  createdAt: Date;
};

export function articleIsPhoto(content: string): boolean {
  return content.includes("<img");
}

export function articleRotation(content: string): number | null {
  if (!articleIsPhoto(content)) return null;
  const m = content.match(/<img src="(.+?)" data-rotation="([^"]+)"/);
  if (!m) return null;
  const r = Number(m[2]);
  return Number.isFinite(r) ? r : null;
}

export function fromArticleDoc(doc: QueryDocumentSnapshot<DocumentData>): Article {
  const d = doc.data();
  return {
    id: doc.id,
    content: d.content,
    signature: d.signature,
    left: d.left,
    top: d.top,
    width: d.width,
    createdAt: (d.createdAt as Timestamp).toDate(),
    createdBy: d.createdBy,
  };
}

export function toArticleDoc(input: ArticleInput): DocumentData {
  return {
    content: input.content,
    signature: input.signature,
    left: input.left,
    top: input.top,
    width: input.width,
    createdAt: Timestamp.fromDate(input.createdAt),
    createdBy: input.createdBy,
  };
}
